### ISSUE TITLE
Feature Request: "custom stage” mechanism

### ISSUE THREAD

**Issue by ShotaKitazawa** [pipe-cd/pipecd#1089]
**What would you like to be added**:

"custom stage” mechanism (e.g. https://spinnaker.io/guides/operator/custom-job-stages/)

**Why is this needed**:

Now, PipeCD support some stages. 
(eg. Kubernetes: https://pipecd.dev/docs/user-guide/configuring-deployment/kubernetes/)

But I want to execute some processes (eg. Run fixed K8s Job) only writing in .pipe.yaml

I hope feature of custom (user-definition) stage.

**Comment by nghialv** [comment]
@ShotaKitazawa Thank you for raising the issue. This is definitely a worthwhile feature. We will add it to the roadmap.

**Comment by nghialv** [comment]
We are thinking about adding a `CUSTOM_BIN` stage that executes the specified binary.

**Comment by kevin-namba** [comment]
1. use custom stage as a deployment stage that just once apply infrastructure like `TERRAFORM_SYNC`
- The pipeline is composed of one custom stage named `CUSTOM_SYNC`
```yaml
pipeline:
    stages:
      - name: CUSTOM_SYNC
        desc: "first"
        with:
          runs:
            - "sam build"
            - "echo y | sam deploy"
```
2. use custom stage as deployment stages that supports a canary release.
- The pipeline is composed of some custom stages
```yaml
  pipeline:
    stages:
      - name: CUSTOM_CANARY
        with:
          runs: 
            - "sam build"
            - "echo y | sam deploy"
      - name: CUSTOM_CANARY_PROMOTE
        desc: "50%"
        with:
          runs: 
            - "echo y | sam deploy --parameter-overrides FUNCTION_WAIT=0.5"
      - name: WAIT
        with:
          duration: 1m
      - name: CUSTOM_CANARY_PROMOTE
        desc: "100%"
        with:
          runs: 
            - "echo y | sam deploy --parameter-overrides FUNCTION_WAIT=1"
            - 
3. use custom stage as script running stage such as webhook, E2E test and display something
- The pipeline is composed of some custom stages and stages related with platform provider.
  pipeline:
    stages:
      - name: K8S_CANARY_ROLLOUT
        with:
          replicas: 10%
      - name: COMMAND_EXEC
        desc: "webhook"
        with:
          runs:
            - "curl https://hooks.slack.com/services"
      - name: WAIT
        with:
          duration: 10s
      - name: K8S_PRIMARY_ROLLOUT
      - name: K8S_CANARY_CLEAN

**Comment by kentakozuka** [comment]
This feature will be released in v0.44.0 for only AWS Lambda.
